agent/charms/testflinger-agent-host-charm/src/charm.py

```diff
@@ -126,6 +131,54 @@ def update_config_files(self):
             shutil.rmtree(repo_path, ignore_errors=True)
         shutil.move(tmp_repo_path, repo_path)
 
+    def write_supervisor_service_files(self):
+        """
+        Generate supervisord service files for all agents
+
+        We assume that the path pointed to by the config-dir config option
+        contains a directory for each agent that needs to run from this host.
+        The agent directory name will be used as the service name.
+        """
+
+        config_dirs = Path(AGENT_CONFIGS_PATH) / self.config.get("config-dir")
+
+        if not config_dirs.is_dir():
+            logger.error("config-dir must point to a directory")
+            self.unit.status = BlockedStatus(
+                "config-dir must point to a directory"
+            )
+            sys.exit(1)
+
+        agent_dirs = [dir for dir in config_dirs.iterdir() if dir.is_dir()]
+        if not agent_dirs:
+            logger.error("No agent directories found in config-dirs")
+            self.unit.status = BlockedStatus(
+                "No agent directories found in config-dirs"
+            )
+            sys.exit(1)
+
+        # Remove all the old service files in case agents have been removed
+        for conf_file in os.listdir("/etc/supervisor/conf.d"):
+            if conf_file.endswith(".conf"):
+                os.unlink(f"/etc/supervisor/conf.d/{conf_file}")
```

Comment:
Clearing all the service files and re-writing seems reliable, but I'm wondering if you considered performing a comparison with the new intended configuration and only deleting or updating where necessary? I would see the small potential benefit being that someone inspecting the agent host could see from file creation times whether an agent is new or updated.

Reply:
@jocave This is just for the supervisor service files, which all follow the same template. What changes are you anticipating here that we need to look at the timestamp on them? There will be new ones that show up if we add agents, but you can also see when they were added to the repo.

I suppose we could do something like this, but I'd like to better understand what we're trying to achieve. I saw this as a much simpler approach which also ensures that we remove any agents that are no longer valid.

Reply:
Just trying to think of ideas that might help in managing an agent host that has been running for some period of time. I can't really say what problems might arise, but just thought there could be a situation in which it would be helpful to know which agents are new after an update and which have had instance that (according to the config produced by the charm) should have been running for longer.

Reply:
Not sure if this helps, but you can see how long an agent has been running:
```
$ supervisorctl status
agent-001                        RUNNING   pid 7218, uptime 3 days, 18:05:33
agent-002                        RUNNING   pid 7219, uptime 3 days, 18:05:33
agent-003                        RUNNING   pid 7220, uptime 3 days, 18:05:33
agent-004                        RUNNING   pid 7221, uptime 3 days, 18:05:33
agent-005                        RUNNING   pid 7222, uptime 3 days, 18:05:33
agent-006                        RUNNING   pid 7223, uptime 3 days, 18:05:33
agent-007                        RUNNING   pid 7224, uptime 3 days, 18:05:32
agent-008                        RUNNING   pid 7225, uptime 3 days, 18:05:32
agent-009                        RUNNING   pid 7226, uptime 3 days, 18:05:32
...
```

But there are also certain situations, such as when we update the testflinger code, where we will also trigger ALL the agents to restart when they are not running a job. We also do this for all agents when the configs have been updated. Otherwise, we would have to keep track of any possible change to any config  within the testflinger and device connector configs across all agents. I think it's less error prone to let them all safely restart and ensure they are running the latest config though.

Reply:
resolving so I can land this, per our mm conversaton